base_user_role/models/role.py

```diff
@@ -66,12 +67,27 @@ class ResUsersRoleLine(models.Model):
     _description = 'Users associated to a role'
 
     role_id = fields.Many2one(
-        'res.users.role', string=u"Role", ondelete='cascade')
+        'res.users.role', required=True, string=u"Role", ondelete='cascade')
     user_id = fields.Many2one(
-        'res.users', string=u"User", domain=[('id', '!=', SUPERUSER_ID)])
+        'res.users', required=True, string=u"User",
+        domain=[('id', '!=', SUPERUSER_ID)], ondelete='cascade')
     date_from = fields.Date(u"From")
     date_to = fields.Date(u"To")
     is_enabled = fields.Boolean(u"Enabled", compute='_compute_is_enabled')
+    company_id = fields.Many2one(
+        'res.company', 'Company',
+        default=lambda self: self.env.user.company_id)
+
+    @api.multi
+    @api.constrains('user_id', 'company_id')
+    def _check_company(self):
+        for record in self:
+            if (self.company_id and
```

Comment:
replace "self" by "record" as you are in the loop. Current code will not work 

Maybe the code can be simplify (more explicit) with
allowed_companies = record.user_id.company_id + record.user_id.company_ids
if record.company_id and record.company_id not in allowed_companies:
   raise ...